packages/block-library/src/template-part/edit/placeholder.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * WordPress dependencies
+ */
+import { useEntityBlockEditor, EntityProvider } from '@wordpress/core-data';
+import { __ } from '@wordpress/i18n';
+import { BlockPreview } from '@wordpress/block-editor';
+import { useState, useCallback } from '@wordpress/element';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { cleanForSlug } from '@wordpress/editor';
+import { Placeholder, TextControl, Button } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import useTemplatePartPost from './use-template-part-post';
+
+function TemplatePartPreview() {
+	const [ blocks ] = useEntityBlockEditor( 'postType', 'wp_template_part' );
+	return (
+		<div className="wp-block-template-part__placeholder-preview">
+			<div className="wp-block-template-part__placeholder-preview-title">
+				{ __( 'Preview' ) }
+			</div>
+			<BlockPreview blocks={ blocks } />
+		</div>
+	);
+}
+
+export default function TemplatePartPlaceholder( { setAttributes } ) {
+	const [ slug, _setSlug ] = useState();
+	const [ theme, setTheme ] = useState();
+	const [ help, setHelp ] = useState();
+
+	// Try to find an existing template part.
+	const postId = useTemplatePartPost( null, slug, theme );
+
+	// If found, get its preview.
+	const preview = useSelect(
+		( select ) => {
+			if ( ! postId ) {
+				return;
+			}
+			const templatePart = select( 'core' ).getEntityRecord(
+				'postType',
+				'wp_template_part',
+				postId
+			);
+			if ( templatePart ) {
+				return (
+					<EntityProvider kind="postType" type="wp_template_part" id={ postId }>
+						<TemplatePartPreview />
+					</EntityProvider>
+				);
+			}
+		},
+		[ postId ]
+	);
+
+	const setSlug = useCallback( ( nextSlug ) => {
+		_setSlug( nextSlug );
+		setHelp( cleanForSlug( nextSlug ) );
+	}, [] );
+
+	const { saveEntityRecord } = useDispatch( 'core' );
+	const onChooseOrCreate = useCallback( async () => {
+		const nextAttributes = { slug, theme };
+		if ( postId !== undefined && postId !== null ) {
+			// Existing template part found.
+			nextAttributes.postId = postId;
+		} else {
+			// Create a new template part.
+			try {
+				const cleanSlug = cleanForSlug( slug );
+				const templatePart = await saveEntityRecord(
+					'postType',
+					'wp_template_part',
+					{
+						title: cleanSlug,
+						status: 'publish',
+						slug: cleanSlug,
+						meta: { theme },
+					}
+				);
+				nextAttributes.postId = templatePart.id;
+			} catch ( err ) {
+				setHelp( __( 'Error adding template.' ) );
+			}
+		}
+		setAttributes( nextAttributes );
+	}, [ postId, slug, theme ] );
+	return (
+		<Placeholder
+			icon="layout"
+			label={ __( 'Template Part' ) }
+			instructions={ __(
+				'Choose a template part by slug and theme, or create a new one.'
+			) }
+		>
+			<TextControl
+				label={ __( 'Slug' ) }
+				placeholder={ __( 'header' ) }
+				value={ slug }
+				onChange={ setSlug }
+				help={ help }
+				className="wp-block-template-part__placeholder-input"
+			/>
+			<TextControl
```

Comment:
It's a bit odd to have this exposed here as an editable attribute since there is already "current theme" context that can be implicit.

Reply:
Yes, but you still need to be able to add template parts from non-active themes that you installed and customized.

I was thinking of filling it with the current theme by default, and maybe hiding it behind some sort of "advanced" checkbox/option?

What do you think?

Reply:
Yes, pre-filling and hiding by default sounds good. The mixing is a bit more complex and should not be immediately exposed.